### PR TITLE
Define algorithm providers in terms of plugins

### LINKED ISSUE #86162: Define algorithm providers in terms of plugins instead of predicates/priorities [Migration Phase 2]
/assign @ahg-g @draveness 
/sig scheduling
/priority important-soon

Part of #85822

### PR DESCRIPTION
**Pull request by ahg-g**
**What type of PR is this?**

/kind cleanup
/kind feature

**What this PR does / why we need it**:
Define algorithm providers in terms of plugins rather than the legacy predicates/priorities keys. This practically makes the scheduler plugins-first instead of predicates/priorities-first.

This PR also defines a new algorithm registry named LegacyRegistry that maintains the old set of registered predicates and priorities. It is the same as the existing one that used to be defined as a set of global variables in ```pkg/scheduler/algorithm_factory.go```. The second commit merges ```ConfigProducerRegistry``` into ```LegacyRegistry``` (instead of being a member of it), and so  we end up with only three registries:

1) A ```framework.Registry``` for in-tree framework plugins defined in ```pkg/scheduler/framework/plugins/registry.go```.

2) ```plugins.LegacyRegistry``` defined in ```pkg/scheduler/framework/plugins/legacy_registry.go```, and contains predicates and priorities keys and how they map to plugins.

3) ```algorithmprovider.Registry``` defined in ```pkg/scheduler/algorithmprovider/registry.go```. A registry of algorithm providers (e.g., DefaultAlgorithmProvider, ClusterAutoscalerProvider) to their plugin configurations.

**Which issue(s) this PR fixes**:
Fixes #86162

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/cc @Huang-Wei

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -250,22 +229,14 @@ var defaultSchedulerOptions = schedulerOptions{
...
 		Provider: defaultAlgorithmSourceProviderName(),
 	},
-	hardPodAffinitySymmetricWeight:  v1.DefaultHardPodAffinitySymmetricWeight,
-	disablePreemption:               false,
-	percentageOfNodesToScore:        schedulerapi.DefaultPercentageOfNodesToScore,
-	bindTimeoutSeconds:              BindTimeoutSeconds,
-	podInitialBackoffSeconds:        int64(internalqueue.DefaultPodInitialBackoffDuration.Seconds()),
-	podMaxBackoffSeconds:            int64(internalqueue.DefaultPodMaxBackoffDuration.Seconds()),
-	frameworkConfigProducerRegistry: frameworkplugins.NewConfigProducerRegistry(),
-	// The plugins and pluginConfig options are currently nil because we currently don't have
```
Comment (liu-cong) [inline pkg/scheduler/scheduler.go]: Is the comment still applicable?

File: pkg/scheduler/apis/config/testing/compatibility_test.go [hunk 2]
```diff
@@ -1363,3 +1365,167 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
+	// Add serialized versions of scheduler config that exercise available options to ensure compatibility between releases
+	testcases := []struct {
+		name        string
+		provider    string
+		wantPlugins map[string][]config.Plugin
+	}{
+		{
+			name: "No Provider specified",
+			wantPlugins: map[string][]config.Plugin{
+				"PreFilterPlugin": {
```
Comment (liu-cong) [inline pkg/scheduler/apis/config/testing/compatibility_test.go]: nit: maybe define a `wantDefaultPluigns` variable to reflect that "No Provider" or "Default Provider" should both return the `wantDefaultPluigns` .

File: pkg/scheduler/factory.go [hunk 3]
```diff
@@ -210,143 +280,65 @@ func (c *Configurator) CreateFromConfig(policy schedulerapi.Policy) (*Scheduler,
...
 	var plugins schedulerapi.Plugins
 	var pluginConfig []schedulerapi.PluginConfig
-	frameworkConfigProducers := c.pluginConfigProducerRegistry.PriorityToConfigProducer
-	for _, p := range keys.List() {
-		weight, exist := algorithmRegistry.priorityKeys[p]
+	var sortedKeys []string
+	for k := range keys {
+		sortedKeys = append(sortedKeys, k)
+	}
+	sort.Strings(sortedKeys)
```
Comment (liu-cong) [inline pkg/scheduler/factory.go]: Why sorting? Can you add a comment?

Comment (ahg-g) [reply pkg/scheduler/factory.go]: done.

File: pkg/scheduler/apis/config/testing/compatibility_test.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/apis/config/testing/compatibility_test.go]: done.

File: pkg/scheduler/scheduler.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/scheduler.go]: nope, removed.

File: pkg/scheduler/factory.go [hunk 4]
```diff
@@ -122,15 +120,95 @@ func (c *Configurator) Create() (*Scheduler, error) {
...
+	if !exist {
+		return nil, fmt.Errorf("algorithm provider %q is not registered", providerName)
+	}
+
+	// Combine the provided plugins with the ones from component config.
+	// TODO(#86162): address disabled plugins.
+	var plugins schedulerapi.Plugins
+	plugins.Append(provider.FrameworkPlugins)
+	plugins.Append(c.plugins)
+	c.plugins = &plugins
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/factory.go]: yes, the order is specified here: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/types.go#L158

We still need to address disabled plugins, which will allow for reordering.

File: pkg/scheduler/factory.go [hunk 5]
```diff
@@ -122,15 +120,95 @@ func (c *Configurator) Create() (*Scheduler, error) {
...
+	var pluginConfig []schedulerapi.PluginConfig
+	pluginConfig = append(pluginConfig, provider.FrameworkPluginConfig...)
+	pluginConfig = append(pluginConfig, c.pluginConfig...)
+	c.pluginConfig = pluginConfig
+
+	return c.CreateFromPlugins([]algorithm.SchedulerExtender{})
+}
+
+// CreateFromPlugins creates a scheduler from a set of registered plugins.
+func (c *Configurator) CreateFromPlugins(extenders []algorithm.SchedulerExtender) (*Scheduler, error) {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/factory.go]: done.

**Comment by ahg-g** [comment]
Thanks @liu-cong, comments addressed.

File: pkg/scheduler/factory.go [hunk 6]
```diff
@@ -112,36 +112,8 @@ type Configurator struct {
...
-	pluginConfig = append(pluginConfig, provider.FrameworkPluginConfig...)
-	pluginConfig = append(pluginConfig, c.pluginConfig...)
-	c.pluginConfig = pluginConfig
-
-	return c.CreateFromPlugins([]algorithm.SchedulerExtender{})
-}
-
-// CreateFromPlugins creates a scheduler from a set of registered plugins.
-func (c *Configurator) CreateFromPlugins(extenders []algorithm.SchedulerExtender) (*Scheduler, error) {
+// createFromPlugins creates a scheduler from a set of registered plugins.
```
Comment (liu-cong) [inline pkg/scheduler/factory.go]: nit: `createFromPlugins` -> `create`

Comment (ahg-g) [reply pkg/scheduler/factory.go]: done. I will squash.

File: pkg/scheduler/scheduler.go (hunk 1 above)
Comment (liu-cong) [reply pkg/scheduler/scheduler.go]: The previous comment says the parameters should be populated once we have native plugins. Can you populate the parameters now?

Comment (ahg-g) [reply pkg/scheduler/scheduler.go]: This wasn't the right place to populate them, the new comment clarifies that those parameters only carry the plugins specified in Component Config.

Comment (ahg-g) [reply pkg/scheduler/scheduler.go]: The default plugins are now populated when calling "createFromProvider".

Comment (liu-cong) [reply pkg/scheduler/scheduler.go]: Ack. Can we just remove these two lines (using go default values) then?

Comment (ahg-g) [reply pkg/scheduler/scheduler.go]: done.

File: pkg/scheduler/apis/config/testing/compatibility_test.go [hunk 7]
```diff
@@ -1312,7 +1314,7 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 			for feature, value := range tc.featureGates {
 				defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, feature, value)()
 			}
-			defer algorithmprovider.ApplyFeatureGates()()
```
Comment (Huang-Wei) [inline pkg/scheduler/apis/config/testing/compatibility_test.go]: Let's also cleanup other occurrences in `test/integration`:

```bash
⇒  ag "ApplyFeatureGates" test/integration/scheduler
test/integration/scheduler/priorities_test.go
251:	defer algorithmprovider.ApplyFeatureGates()()

test/integration/scheduler/taint_test.go
87:	defer algorithmprovider.ApplyFeatureGates()()
651:	// ApplyFeatureGates() is called to ensure TaintNodesByCondition related logic is applied/restored properly.
652:	defer algorithmprovider.ApplyFeatureGates()()

test/integration/scheduler/predicates_test.go
883:	defer algorithmprovider.ApplyFeatureGates()()
```

File: pkg/scheduler/algorithmprovider/registry_test.go [hunk 8]
```diff
@@ -17,83 +17,163 @@ limitations under the License.
...
-				config.FrameworkPluginConfig,
-				framework.WithClientSet(client),
-				framework.WithInformerFactory(informerFactory),
-				framework.WithSnapshotSharedLister(nodeinfosnapshot.NewEmptySnapshot()),
-			)
-			if err != nil {
-				t.Fatalf("error initializing the scheduling framework: %v", err)
-			}
-			wantPlugins := fwk.ListPlugins()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ResourceLimitsPriorityFunction, true)()
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithmprovider/registry_test.go]: Can we add a case to ensure PodTopologySpread plugins are not enabled when the featuregate is disabled?

File: pkg/scheduler/factory.go [hunk 9]
```diff
@@ -210,143 +275,68 @@ func (c *Configurator) CreateFromConfig(policy schedulerapi.Policy) (*Scheduler,
...
 		return nil, err
 	}
-
 	// Combine all framework configurations. If this results in any duplication, framework
 	// instantiation should fail.
 	var plugins schedulerapi.Plugins
 	plugins.Append(pluginsForPredicates)
 	plugins.Append(pluginsForPriorities)
 	plugins.Append(c.plugins)
+	c.plugins = &plugins
```
Comment (Huang-Wei) [inline pkg/scheduler/factory.go]: It seems `c.plugins` have to be appended again to itself, how about:

```go
if c.plugins == nil {
	c.plugins = &schedulerapi.Plugins{}
}
c.plugins.Append(pluginsForPredicates)
c.plugins.Append(pluginsForPriorities)
```

File: pkg/scheduler/factory.go [hunk 10]
```diff
@@ -210,143 +275,68 @@ func (c *Configurator) CreateFromConfig(policy schedulerapi.Policy) (*Scheduler,
...
 	var pluginConfig []schedulerapi.PluginConfig
 	pluginConfig = append(pluginConfig, pluginConfigForPredicates...)
 	pluginConfig = append(pluginConfig, pluginConfigForPriorities...)
 	pluginConfig = append(pluginConfig, c.pluginConfig...)
+	c.pluginConfig = pluginConfig
```
Comment (Huang-Wei) [inline pkg/scheduler/factory.go]: Ditto.

**Review by Huang-Wei — COMMENTED** [review]
Thanks @ahg-g . Some nits below.

File: pkg/scheduler/factory.go (hunk 9 above)
Comment (ahg-g) [reply pkg/scheduler/factory.go]: This doesn't work because order matter, we need it to be last.

File: pkg/scheduler/factory.go (hunk 10 above)
Comment (ahg-g) [reply pkg/scheduler/factory.go]: addressed above

File: pkg/scheduler/algorithmprovider/registry_test.go (hunk 8 above)
Comment (ahg-g) [reply pkg/scheduler/algorithmprovider/registry_test.go]: done.

File: pkg/scheduler/apis/config/testing/compatibility_test.go (hunk 7 above)
Comment (ahg-g) [reply pkg/scheduler/apis/config/testing/compatibility_test.go]: done.

**Pull request merged**